clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -848,6 +854,24 @@ private TxnOffsetCommitHandler txnOffsetCommitHandler(TransactionalRequestResult
         return new TxnOffsetCommitHandler(result, builder);
     }
 
+    private TransactionalRequestResult handleCachedTransactionRequestResult(
+            Supplier<TransactionalRequestResult> transactionalRequestResultSupplier,
+            State targetState) {
+        ensureTransactional();
+
+        // null it out if the cached result was already completed
+        if (transactionalRequestResult != null && transactionalRequestResult.isCompleted()) {
```

Comment:
We need to make sure that the user observes the completion. If the state is correct, we should return the current `TransactionalRequestResult` and set the cached value to null. I would suggest we get rid of this check and do something like this:

```java
if (transactionalRequestResult != null) {
  if (currentState != targetState)
    throw new IllegalStateException()

  TransactionalRequestResult result = transactionalRequestResult;
  if (result.isCompleted())
    transactionalRequestResult = null;
  return result;
}
```

Does that make sense?

Reply:
@hachikuji  Thanks for all  the comments. If the code does not retry, the previous failed operation leaves the cached result completed but also illegal to the next transaction operation. In this case, an IllegalStateException is thrown. So I think it's a good idea to clear the result at  the very beginning of the method.

Reply:
If the user does not observe the completion, then how will they know it completed successfully? Let me give a specific scenario. Let's say the user calls `commitTransaction()` and it times out. Just after timing out, the operation returns successfully. Now what happens when the user calls `commitTransaction()`? What should happen is that we notice that the pending operation has completed and we return immediately.